Create a tests view for test workspaces

jaril opened the pull request:
This adds a tests view, as detailed [here](https://www.notion.so/Testsuites-V2-Design-Spec-be3f0201e26e494090ca6b9e62fbd26e). This is for displaying all the tests recorded for a workspace, and surfacing useful information like failure rates over time, passing replays, and failing replays.

In a way, this is not complete yet. We're not able to query performantly yet because some of the data that this view was designed to handle. So added this under a feature flag for now until we're able to turn it on. To enable the test view, go to the console and run `localStorage.setItem("enableTestSuitesTestsView", true)`.

Other notes:
- Failure rates are currently calculated in the frontend. This will be moved to the backend eventually
- We have a time filter that should apply to both 1) tests and 2) errors. That API's not implemented yet so the corresponding dropdowns in the frontend don't do anything right now
- There's a sort by feature for tests as well. Same as above, it's not implemented yet, so the dropdown doesn't do anything for now

jasonLaster commented:
1. Added a fixture so that data would load quickly 
2. Copied a bunch of the TestRun files into the Tests directory
3. Made a bunch of renaming fixes

jaril commented:
1. Fixed displaying the test list
2. Pre-computed the failure rate in `src/ui/components/Library/Team/View/TestRuns/suspense/TestsCache.ts`
3. Started displaying the failure rates in the right hand side panel

bvaughn commented:
What is this PR? No description and a huge delta.

<img width="155" alt="Screen Shot 2023-11-14 at 10 59 14 AM" src="https://github.com/replayio/devtools/assets/29597/463696f2-befa-4ef6-aaf4-bd25d0ea79bc">

bvaughn commented on src/ui/components/Library/Team/View/Tests/TestContextRoot.tsx:
```diff
@@ -0,0 +1,95 @@
+
+    return filteredTestRuns;
+  }, [filterByText, tests]);
+  // }, [filterByBranch, filterByStatus, filterByText, testRuns]);
+
+  // useEffect(() => {
+  //   if (testRunId == null) {
+  //     // Select the first test run by default if nothing is selected.
+  //     setTestRunId(testRuns[0]?.id ?? null);
+  //   }
+  // }, [router, teamId, testRunId, testRuns]);
+
+  // useSyncTestRunIdToUrl(teamId, testRunId, setTestRunId);
```
Kill the dead code:
```suggestion
```

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestOverviewContainerContextType.tsx:
```diff
@@ -0,0 +1,8 @@
+
+export const TestOverviewContext = createContext<TestOverviewContainerContextType>(null as any);
+
+export function TestOverviewContainer({ children }: { children: ReactNode }) {
+  return <TestOverviewContext.Provider value={{}}>{children}</TestOverviewContext.Provider>;
+}
```
This component doesn't really add much value. Since it's only used in one place (`TestOverviewPage`) can you just inline the context provider there?

jaril replied:
Yep — deleted this structure so we can just use the `TestOverviewContent` component directly

bvaughn commented on src/ui/components/Library/Team/View/Tests/hooks/useTests.ts:
```diff
@@ -0,0 +1,25 @@
+
+const EMPTY_ARRAY: any[] = [];
+
+export function useTests(): any[] {
```
Replace `any[]` with an actual type.

`useTestRuns()` returns a `TestRun[]`. This should probably return a `Test[]` array (presumably the `Test` type that you added to `packages/shared/test-suites/TestRun`)

bvaughn commented on src/ui/components/Library/Navigation/TeamButton.tsx:
```diff
@@ -43,25 +45,73 @@ export function TeamButton({
+    <div className="pl-4 pb-2">
+      <div className="flex flex-col">
+        <div
+          className={`px-4 py-1 ${
```
The margin and alignment for this sub menu looks awkward to me. The "Runs" text is not indented enough to look "nested" beneath the team name:

<img width="266" alt="Screen Shot 2023-11-15 at 10 20 58 AM" src="https://github.com/replayio/devtools/assets/29597/86abb840-f294-4cea-bb6a-2951490daf83">

And it looks lopsided vertically because of the added bottom padding:

<img width="269" alt="Screen Shot 2023-11-15 at 10 21 12 AM" src="https://github.com/replayio/devtools/assets/29597/cc2bee3f-9942-4577-864a-4169510ae351">

I suggest changing it like so for now:
```suggestion
          className={`px-8 ${
```

<img width="260" alt="Screen Shot 2023-11-15 at 10 22 55 AM" src="https://github.com/replayio/devtools/assets/29597/752a8f98-8bb7-4c98-98f0-23d6bf4e6f51">

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestErrorList.tsx:
```diff
@@ -0,0 +1,176 @@
+import { TestExecution } from "shared/test-suites/TestRun";
+
+import { getTruncatedRelativeDate } from "../../Recordings/RecordingListItem/RecordingListItem";
+import styles from "../../../../Library.module.css";
+import testPageStyles from "../TestsPage.module.css";
```
This feels kind of clunky. Seems like maybe the "drop down trigger" thing could be its own reusable component rather than having views import each others styles like this.

In general, I don't like that pattern because I think it's easy to make (what you think is) a localized style edit that breaks something else (because you don't assume that other components import your style modules).

bvaughn commented on src/ui/components/Library/Navigation/TeamButton.tsx:
```diff
@@ -43,25 +45,73 @@ export function TeamButton({
+          } hover:cursor-pointer hover:font-bold`}
+          onClick={() => setView("runs")}
+        >
+          Runs
```
We should only be showing this sub-view for tests suites pages, right? This PR shows it for everything:
https://www.loom.com/share/52299129b46e4430962fac5396e7f588

Also I think we should only show the submenus ("Runs" and "Tests") if the `enableTestView` feature is on, but this PR shows it regardless which is kind of confusing. (Why show only a single option?)

bvaughn replied:
This sub-menu is also visible under "Your Library"
<img width="538" alt="Screen Shot 2023-11-15 at 10 44 37 AM" src="https://github.com/replayio/devtools/assets/29597/4fa3b61b-a156-4f45-a0af-5ff90ea536ce">

And clicking these just shows an empty page (because this isn't a test suite workspace).

bvaughn commented on src/ui/components/Library/Team/View/TestRuns/suspense/TestsCache.ts:
```diff
@@ -0,0 +1,79 @@
+  load: async ([graphQLClient, accessToken, workspaceId]) => {
+    const rawTests = await getTestsGraphQL(graphQLClient, accessToken, workspaceId);
+
+    // Compute the failure rate and add it to the stored test in cache
+    // until we're able to do it from the backend
```
Comments like this are easy to overlook and forget about. Add a TODO that links to a Linear issue for this?

jaril replied:
Sounds good — captured in https://linear.app/replay/issue/SCS-1575/compute-the-failure-rates-in-the-backend-instead-of-the-frontend

bvaughn commented on src/ui/components/Library/Navigation/TeamButton.tsx:
```diff
@@ -43,25 +45,73 @@ export function TeamButton({
+        <div
+          className={`px-4 py-1 ${
+            view === "runs" ? "font-bold" : ""
+          } hover:cursor-pointer hover:font-bold`}
```
I think we should probably also avoid using "bold" style for hover state. Having text resize on mouse over feels jarring.

```suggestion
          } hover:cursor-pointer`}
```

bvaughn commented on packages/shared/user-data/LocalStorage/config.ts:
```diff
@@ -10,4 +10,5 @@ export const config = {
   reactDevToolsStateExpanded: Boolean(true),
   replayNextCurrentPanel: "sources" as ReplayNextCurrentPanel,
   replayVideoPanelCollapsed: false,
+  enableTestView: false,
```
nit: This could use a more descriptive name. Maybe something like `enableTestSuitesTestsView` or something?

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestErrorList.tsx:
```diff
@@ -0,0 +1,176 @@
+  const [selectedError, setSelectedError] = useState<string | null>(null);
+  const [filterByTime, setFilterByTime] = useState<number | null>(null);
+
+  const {
+    contextMenu: contextMenuTimeFilter,
+    onContextMenu: onClickTimeFilter,
+    onKeyDown: onKeyDownTimeFilter,
+  } = useContextMenu(
+    <>
+      <ContextMenuItem dataTestId="show-all-runs" onSelect={() => setFilterByTime(null)}>
+        All errors
+      </ContextMenuItem>
+    </>,
+    { alignTo: "auto-target" }
```
Only adding one option to these menus makes them kind of useless:

<img width="174" alt="Screen Shot 2023-11-15 at 10 46 55 AM" src="https://github.com/replayio/devtools/assets/29597/24294838-1b44-48c6-bf5d-95afcdb95c9e">

<img width="106" alt="Screen Shot 2023-11-15 at 10 46 58 AM" src="https://github.com/replayio/devtools/assets/29597/4ffc9daf-7456-49df-9444-8faa887a74b7">

<img width="120" alt="Screen Shot 2023-11-15 at 10 47 02 AM" src="https://github.com/replayio/devtools/assets/29597/1c46da8c-13c6-45a6-a4d1-581f2016dd80">

There's no way to turn them off.

jasonLaster replied:
Lets show the other options that are in the mock, but have them be disabled. This will let users see that they're coming

bvaughn reviewed (commented):
Initial batch of feedback.

bvaughn commented:
@jaril TypeScript is failing in CI: https://github.com/replayio/devtools/actions/runs/6882322396/job/18720591892?pr=9892

jonbell-lot23 reviewed (commented):
1. The capsule text should be #fff
2. Please add  percentage symbol to the end: "50%" rather than "50"

<img width="170" alt="image" src="https://github.com/replayio/devtools/assets/9154902/543158b0-67af-4e6c-9a0c-dc8f234aa965">


4. When the right column is empty, it should have a little note like this: 

<img width="379" alt="image" src="https://github.com/replayio/devtools/assets/9154902/c3934072-fb0c-4d46-a546-c07902e88cfd">

5. Why are these replays dimmed? 


<img width="333" alt="image" src="https://github.com/replayio/devtools/assets/9154902/1bb51761-c547-4c25-b5bf-349ec23d941c">

6. 

We shouldn't show a header when there's no content below it

<img width="349" alt="image" src="https://github.com/replayio/devtools/assets/9154902/66d581e6-38af-4d28-9e00-32a050b1059e">


7. If something has never failed, I guess we should make it green instead of red: 

<img width="490" alt="image" src="https://github.com/replayio/devtools/assets/9154902/5c9ba33c-46d3-47d1-8d88-c427fe267e21">

I don't think any of these are strictly required to push live. Happy to handle these on the design polish side so you can focus on the nuts and bolts.

jaril commented:
> 1. The capsule text should be #fff
Fixed!

> 2. Please add  percentage symbol to the end: "50%" rather than "50"
Fixed!

> 3. When the right column is empty, it should have a little note like this:
Good catch. Adding.

> 4. Why are these replays dimmed?
The backend doesn't pass those replays to the frontend correctly yet, so we know when they were executed but we don't know what their IDs are. It should be fixed soon:tm:

> 6. If something has never failed, I guess we should make it green instead of red:
Copy!

bvaughn commented on src/ui/components/Library/LibraryDropdownTrigger.tsx:
```diff
@@ -0,0 +1,32 @@
+  testId?: string;
+}) {
+  return (
+    <>
```
nit: Remove the `<>` (fragment) wrapper. It's unnecessary.

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestErrorList.tsx:
```diff
@@ -0,0 +1,181 @@
+  const sortedFailing = orderBy(failing, "createdAt", "desc");
+  const sortedPassing = orderBy(passing, "createdAt", "desc");
+
+  // TODO: show the processed/unprocessed status for the replays -jvv
+  // TODO: add the top commit for the corresponding replay as label -jvv
```
nit: Nice to link these to a Linear issue (e.g. `// TODO [SCS-#]`)

bvaughn commented on src/ui/components/Library/Navigation/TeamButton.tsx:
```diff
@@ -43,25 +46,62 @@ export function TeamButton({
+          </div>
+        </span>
+        {isNew ? (
+          <div className={"rounded-md bg-primaryAccent px-3 py-0.5 text-xs text-white"}>New</div>
```
This is a pretty wide button. Testing it I realized that we don't really handle text overflow well in the side nav:
<img width="264" alt="Screen Shot 2023-11-16 at 8 10 34 AM" src="https://github.com/replayio/devtools/assets/29597/83463949-cc5f-474b-9677-124f4fb14373">

I think the above style was meant to address that, but doesn't:
```html
<span className="overflow-hidden overflow-ellipsis whitespace-pre">
```

I think to fix it we'd want to move that wrapper around the `{label}`:
```jsx
<div className="flex overflow-hidden">
  <LibraryIcon teamType={isTest ? "tests" : "team"} />
  <span className="overflow-hidden overflow-ellipsis whitespace-pre">{label}</span>
</div>
```

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestErrorList.tsx:
```diff
@@ -0,0 +1,181 @@
+      <ContextMenuItem disabled onSelect={() => setFilterByTime(30)}>
+        Last month
+      </ContextMenuItem>
+      <ContextMenuItem onSelect={() => setFilterByTime(null)}>All time</ContextMenuItem>
```
There's something that feels off to me about showing time filters for both the outer test runs column _and_ the inner top errors column. For example, filtering the outer test runs column to show only failed tests for the past hour and then filtering top errors for the selected test to be for "all time" feels misleading and confusing. Seems like the inner column should mirror the outer setting?

cc @jonbell-lot23 because I realize this is more of a UX comment.

jonbell-lot23 replied:
Yeah this is a tough one. The middle mirroring the left can cause other issues. I agree that this can be confusing, but sometimes this granularity is needed and I think the alternative is worse. Will be keeping an eye on this for sure.

jaril replied:
@jonbell-lot23 for that one, in which cases do you imagine the second dropdown will be used?

The more I think about this the less sure I am that we need the second time dropdown since it's redundant. Any narrowing-down action that applies to the second dropdown can just as easily be applied with the first dropdown.

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestErrorList.tsx:
```diff
@@ -0,0 +1,181 @@
+          Recent replays of the test passing
+        </div>
+        <div className="flex flex-col gap-2">
+          {sortedPassing.slice(0, MAX_REPLAYS_SHOWN).map((e, i) => (
+            <ErrorReplayList date={e.createdAt} index={i + 1} key={i} status="passed" />
+          ))}
```
We should probably show something for each of these recent blocks when they're empty. This looks kind of broken to me:
<img width="1374" alt="Screen Shot 2023-11-16 at 8 17 26 AM" src="https://github.com/replayio/devtools/assets/29597/9c78cc1b-4596-4614-a694-4f10d46d158d">

jaril replied:
Recently changed this behavior — for now, I just stopped displaying a block if there are no applicable replays, ie "Recent replays of the test passing" wouldn't show up in the first place.

bvaughn commented on src/ui/components/Library/Team/View/Tests/TestContextRoot.tsx:
```diff
@@ -0,0 +1,85 @@
+
+  const deferredTestId = useDeferredValue(testId);
+
+  return (
+    <TestContext.Provider
+      value={{
+        filterByTime,
+        sortBy,
+        filterByText: filterByTextDeferred,
+        filterByTextForDisplay: filterByText,
+        selectTestId: setTestId,
+        setFilterByTime,
+        setSortBy,
+        setFilterByText,
+        testId: deferredTestId,
+        testIdForDisplay: testId,
+        tests: filteredTests,
+      }}
+    >
```
nit: Might be worth memoizing this context value to avoid scheduling unnecessary client updates, but it's probably not necessary.

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestErrorList.tsx:
```diff
@@ -0,0 +1,183 @@
+  const uniqueErrorCount = Object.entries(errorFrequency).length;
+
+  return (
+    <div>
```
This needs overflow. As it is now, if the page is too short to show them all, you can't scroll to see what's cut off. (Test by making your browser page shorter.)
```suggestion
        <div class="overflow-y-auto">
```

bvaughn commented on src/ui/components/Library/Team/View/Tests/Overview/TestOverviewContent.tsx:
```diff
@@ -0,0 +1,77 @@
+
+  return (
+    <div className="border-b border-themeBorder py-2 px-4">
+      <div>
```
Pretty sure this needs to be overflow as well
```suggestion
<div class="overflow-y-auto">
```

jaril commented:
@bvaughn Ready for another look! 

Note there seems to be delta failures, but I'm not sure they're related.

jaril commented:
@bvaughn Ready for another look!

bvaughn reviewed (commented):
Vertical overflow still doesn't scroll...

jaril commented:
> Vertical overflow still doesn't scroll...

My bad — fixing it now.

jaril commented:
Double checked and we should be good with the scrolling!

https://github.com/replayio/devtools/assets/15959269/2f3958c1-dcb0-4fda-81c5-3b5a50a81bc7

bvaughn commented:
RE scrolling: https://www.loom.com/share/f7c6617c7ea74f7889ff2e7462f38b3f cc @jonbell-lot23
